src/StmtToViz.cpp

```diff
@@ -0,0 +1,2900 @@
+
+    void print(const Expr &ir) {
+        ir.accept(this);
+    }
+
+    void print(const Stmt &ir) {
+        ir.accept(this);
+    }
+
+    /* Methods used to emit common HTML patterns */
+
+    // Prints the opening tag for the specified html element. A unique ID is
+    // auto-generated unless provided.
+    void print_opening_tag(const string &tag, const string &cls, int id = -1) {
+        stream << "<" << tag << " class='" << cls << "' id='";
+        if (id == -1) {
+            stream << context_stack.back() << "-";
+            stream << gen_unique_id();
+        } else {
+            stream << id;
+        }
+        stream << "'>";
+        context_stack.push_back(gen_unique_id());
+        context_stack_tags.push_back(tag);
+    }
+
+    void print_opening_tag(const string &tag, const string &cls, string id) {
+        stream << "<" << tag << " class='" << cls << "' id='" << id << "'>";
+        context_stack.push_back(gen_unique_id());
+        context_stack_tags.push_back(tag);
+    }
+
+    // Prints the closing tag for the specified html element.
+    void print_closing_tag(const string &tag) {
+        internal_assert(!context_stack.empty() && tag == context_stack_tags.back())
+            << tag << " " << context_stack.empty() << " " << context_stack_tags.back();
+        context_stack.pop_back();
+        context_stack_tags.pop_back();
+        stream << "</" + tag + ">";
+    }
+
+    // Prints an html element: opening tag, body and closing tag
+    void print_html_element(const string &tag, const string &cls, const string &body, int id = -1) {
+        print_opening_tag(tag, cls, id);
+        stream << body;
+        print_closing_tag(tag);
+    }
+
+    void print_html_element(const string &tag, const string &cls, const string &body, string id) {
+        print_opening_tag(tag, cls, id);
+        stream << body;
+        print_closing_tag(tag);
+    }
+
+    // Prints the opening/closing tags for an anchor that toggles code block view
+    void print_toggle_anchor_opening_tag(int id) {
+        stream << "<a onclick='return toggle(" << id << ");' href=_blank>";
+    }
+
+    void print_toggle_anchor_closing_tag() {
+        stream << "</a>";
+    }
+
+    // Prints a variable to stream
+    void print_variable(const string &x) {
+        stream << variable(x);
+    }
+
+    string variable(const string &x) {
+        int id;
+        if (scope.contains(x)) {
+            id = scope.get(x);
+        } else {
+            id = gen_unique_id();
+            scope.push(x, id);
+        }
+        ostringstream s;
+        s << "<b class='variable matched' id='" << id << "-" << gen_unique_id() << "'>";
+        s << x;
+        s << "</b>";
+        return s.str();
+    }
+
+    // Prints text to stream
+    void print_text(const string &x) {
+        stream << x;
+    }
+
+    // Prints the button to show or hide a code scope
+    void print_show_hide_icon(int id) {
+        stream << "<div class='show-hide-btn-wrapper'>"
```

Comment:
Here and a few other places: is there any risk that the HTML snippets we inject could get out of sync with the template files in a meaningful way?

Reply:
The template files define the javascript and css, whereas what we inject inside `StmtToViz.cpp` is html. Changing the html can definitely break the page as selectors may not match and the buttons will consequently stop working or the styling will not be applied.

Perhaps we can reduce the risk by adding documentation near sensitive code. Let users know that changes to these snippets usually require corresponding changes to the JS.